quote-book/synchronized/src/lib.rs

```diff
@@ -1,111 +1,392 @@
+    }
+}
+#[derive(Clone)]
+/// A wrapper for a quote book implementation that syncs quotes with the ledger
+pub struct SynchronizedQuoteBook<Q: QuoteBook, L: Ledger + Clone + Sync + 'static> {
+    /// Quotebook being synchronized to the ledger by the Synchronized Quotebook
+    synchronized_quote_book_impl: Arc<SynchronizedQuoteBookImpl<Q, L>>,
+}
+impl<Q: QuoteBook, L: Ledger + Clone + Sync + 'static> SynchronizedQuoteBook<Q, L> {
+    /// Create a new Synchronized Quotebook
+    pub fn new(
+        quote_book: Q,
+        ledger: L,
+        remove_quote_callback: RemoveQuoteCallback,
+        logger: Logger,
+    ) -> Self {
+        let synchronized_quote_book_impl = Arc::new(SynchronizedQuoteBookImpl::new(
+            quote_book,
+            ledger,
+            remove_quote_callback,
+            logger,
+        ));
+        Self {
+            synchronized_quote_book_impl,
+        }
+    }
+
+    pub fn get_current_block_index(&self) -> u64 {
+        self.synchronized_quote_book_impl.get_current_block_index()
+    }
+}
+
+impl<Q, L> QuoteBook for SynchronizedQuoteBook<Q, L>
+where
+    Q: QuoteBook,
+    L: Ledger + Clone + Sync + 'static,
+{
+    fn add_sci(
+        &self,
+        sci: SignedContingentInput,
+        timestamp: Option<u64>,
+    ) -> Result<Quote, QuoteBookError> {
+        self.synchronized_quote_book_impl.add_sci(sci, timestamp)
+    }
+
+    fn remove_quote_by_id(&self, id: &QuoteId) -> Result<Quote, QuoteBookError> {
+        self.synchronized_quote_book_impl.remove_quote_by_id(id)
+    }
+
+    fn remove_quotes_by_key_image(
+        &self,
+        key_image: &KeyImage,
+    ) -> Result<Vec<Quote>, QuoteBookError> {
+        self.synchronized_quote_book_impl
+            .remove_quotes_by_key_image(key_image)
+    }
+
+    fn remove_quotes_by_tombstone_block(
+        &self,
+        current_block_index: BlockIndex,
+    ) -> Result<Vec<Quote>, QuoteBookError> {
+        self.synchronized_quote_book_impl
+            .remove_quotes_by_tombstone_block(current_block_index)
+    }
+
+    fn get_quotes(
+        &self,
+        pair: &Pair,
+        base_token_quantity: impl RangeBounds<u64>,
+        limit: usize,
+    ) -> Result<Vec<Quote>, QuoteBookError> {
+        self.synchronized_quote_book_impl
+            .get_quotes(pair, base_token_quantity, limit)
+    }
+
+    fn get_quote_ids(&self, pair: Option<&Pair>) -> Result<Vec<QuoteId>, QuoteBookError> {
+        self.synchronized_quote_book_impl.get_quote_ids(pair)
+    }
+
+    fn get_quote_by_id(&self, id: &QuoteId) -> Result<Option<Quote>, QuoteBookError> {
+        self.synchronized_quote_book_impl.get_quote_by_id(id)
+    }
+
+    fn num_scis(&self) -> Result<u64, QuoteBookError> {
+        self.synchronized_quote_book_impl.num_scis()
+    }
+}
+
+/// A callback for broadcasting a quote removal. It receives 1 argument:
+/// - A vector of quotes that have been removed
+pub type RemoveQuoteCallback = Arc<Mutex<dyn FnMut(Vec<Quote>) + Sync + Send>>;
```

Comment:
Does this need to be `FnMut`? And does it need to be inside a `Mutex`?

Nit: This should move to the top

Reply:
I think so. Previously the calls using the msg_bus_tx seemed to require it.

Reply:
double clicking on this, it seems likely to me that `Arc<Mutex` can be omitted, and it would simplify things if that's the case.